### PR TITLE
[feature] domain block wildcarding

### PR DESCRIPTION
**Pull request by NyaaaWhatsUpDoc**
For domain blocks, it would now also block all subdomains under that given domain.

e.g. if you block `example.com` then the following would also be blocked
- `gts.example.com`
- `pleroma.example.com`
- `gts.pleroma.example.com`

### CODE REVIEW

**Comment by NyaaaWhatsUpDoc** [comment]
Marking as draft as there's still some discussions to be had regarding
- separation of logic, putting this in the database feels like putting too much into the database itself
- how many max subdomain parts is a reasonable amount?

**Comment by NyaaaWhatsUpDoc** [comment]
Thinking about this now... To help separate the logic, once the state PR #1078 has been merged, we could move this logic out of the database quite easily since the main part that holds it here right now is the access to `cache.Has(...)`

**Comment by tsmethurst** [comment]
This looks good so far, very neat implementation :heart_eyes: My only concern is that it adds a lot of database calls (up to three extras, right?) during authorization, which is a very hot path called on every incoming request. I know the cacheing will help a lot but this still makes me a little nervous. What do you think?

**Comment by igalic** [comment]
what about reversing a domain and comparing it?

```
reverse('gts.example.com').startsWith(reverse('example.com') +  '.')
```

shouldn't matter how many domain parts there are then

maybe there's even an `endsWith()` function, so you don't need to reverse it

**Comment by NyaaaWhatsUpDoc** [comment]
> what about reversing a domain and comparing it?
> 
> ```
> reverse('gts.example.com').startsWith(reverse('example.com') +  '.')
> ```
> 
> shouldn't matter how many domain parts there are then
> 
> maybe there's even an `endsWith()` function, so you don't need to reverse it

This is helpful if we have the domain block to compare an incoming domain against. But how do we structure this as part of a database call and a hashmap (cache) lookup?

**Comment by NyaaaWhatsUpDoc** [comment]
> This looks good so far, very neat implementation heart_eyes My only concern is that it adds a lot of database calls (up to three extras, right?) during authorization, which is a very hot path called on every incoming request. I know the cacheing will help a lot but this still makes me a little nervous. What do you think?

I agree, it's definitely a work in progress. If you have any brainwaves on it let me know! I'm going to keep pondering on it myself :thinking:

**Comment by NyaaaWhatsUpDoc** [comment]
> This looks good so far, very neat implementation heart_eyes My only concern is that it adds a lot of database calls (up to three extras, right?) during authorization, which is a very hot path called on every incoming request. I know the cacheing will help a lot but this still makes me a little nervous. What do you think?

I agree, it's definitely a work in progress. If you have any brainwaves on it let me know! I'm going to keep pondering on it myself :thinking:

File: internal/db/bundb/domain.go [hunk 1]
```diff
@@ -82,17 +95,62 @@ func (d *domainDB) CreateDomainBlock(ctx context.Context, block *gtsmodel.Domain
...
+		// Check if this lookup result already cached
+		cached := d.cache.Has("Domain", lookup)
+
+		// Attempt to fetch domain block for lookup
+		block, err := d.getDomainBlock(ctx, lookup)
+		if err != nil {
+			// We return early if this is one of:
+			// - NOT db.ErrNoEntries, i.e. unrecoverable error
+			// - cached db.ErrNoEntries, i.e. have tried this before
+			if !errors.Is(err, db.ErrNoEntries) || cached {
```
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/db/bundb/domain.go:129]: I may have brain-farted this logic. Cached should indeed be checked first :sweat_smile:

**Comment by NyaaaWhatsUpDoc** [comment]
Just had a thought -- the caching side of things could be sped up with a custom lookup system using a radix-trie to do glob-base matching, similar to what I have here: https://codeberg.org/gruf/go-glob/src/branch/main/multi.go.

Still need to figure out the database though.

**Comment by NyaaaWhatsUpDoc** [comment]
I actually came up with quite a nice solution for this, not too complicated but should handle things well :grinning: 

Only thing left is I need to write some tests for the new `domain.BlockCache{}` type!

File: internal/db/bundb/domain.go [hunk 2]
```diff
@@ -50,46 +50,51 @@ func normalizeDomain(domain string) (out string, err error) {
...
 		return nil, err
 	}
 
-	return d.state.Caches.GTS.DomainBlock().Load("Domain", func() (*gtsmodel.DomainBlock, error) {
-		// Check for easy case, domain referencing *us*
-		if domain == "" || domain == config.GetAccountDomain() {
-			return nil, db.ErrNoEntries
-		}
+	// Check for easy case, domain referencing *us*
+	if domain == "" || domain == config.GetAccountDomain() {
```
Comment (tsmethurst) [inline internal/db/bundb/domain.go:82]: Should `domain == config.GetHost()` also be included here? Or is that case handled elsewhere?

**Review by tsmethurst — COMMENTED** [review]
This is looking really nice :)

File: internal/cache/domain/domain.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+}
+
+// block represents a domain block, and stores the
+// deconstructed labels of singular domain block.
+// e.g. []string{"gts", "superseriousbusiness", "org"}.
+type block struct {
+	labels []string
+}
+
+func (b block) Blocks(labels []string) bool {
```
Comment (tsmethurst) [inline internal/cache/domain/domain.go:141]: It's not so easy to understand what's happening here. I think a bit more commenting and/or some examples here would help a lot. Maybe a step-by-step or something? Not necessarily inside the function, but perhaps as a comment on the function?

**Review by tsmethurst — COMMENTED** [review]
Looks great :) Just a comment on the comprehensibility of one of the functions (arguably the most important one) :P

**Pull request merged**